File: iot/amqp/header/impl/AMQPDetach.py
import iot.amqp.avps.AMQPType as AMQPType
import iot.amqp.avps.HeaderCode as HeaderCode
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.tlv.impl.TLVFixed as TLVFixed
import iot.amqp.tlv.impl.TLVList as TLVList
import iot.amqp.tlv.impl.AMQPError as AMQPError

class amqpDetach():
    def __init__(self,code,doff,type,channel,handle,closed,error):
        if code is not None:
            self.code = code
        else:
            self.code = HeaderCode.headerCode.getValueByKey('DETACH')
        if doff is not None:
            self.doff = doff
        else:
            self.doff = 2
        if type is not None:
            self.type = type
        else:
            self.type = 0
        if channel is not None:
            self.channel = channel
        else:
            self.channel = 0
        self.handle = handle
        self.closed = closed
        self.error = error
    
    def setDoff(self,doff):
        self.doff = doff

    def getDoff(self):
        return self.doff

    def getType(self):
        return self.type

    def setType(self, type):
        self.type = type

    def getChannel(self):
        return self.channel

    def setChannel(self, channel):
        self.channel = channel

    def getCode(self):
        return self.code
        
    def toArgumentsList(self):
        list = TLVList.tlvList(None,None)
        wrapper = AMQPWrapper.amqpWrapper()
        if self.handle is None:
            print("Detach header's handle can't be null")
        list.addElement(0, wrapper.wrap(self.handle))
        if self.closed is not None:
            list.addElement(1, wrapper.wrap(self.closed))
        if self.error is not None:
            list.addElement(2, self.error.toArgumentsList())

        constructor = DescribedConstructor.describedConstructor(list.getCode(),TLVFixed.tlvFixed(AMQPType.amqpType.getValueByKey('SMALL_ULONG'), self.code.value))
        list.setConstructor(constructor)
        return list

    def fromArgumentsList(self, list):
        unwrapper = AMQPUnwrapper.amqpUnwrapper()
        size = len(list.getList())
        if size  == 0:
            print("Received malformed Detach header: handle can't be null")
        if size > 3:
            print("Received malformed Detach header. Invalid number of arguments: " + str(size))
        if size > 0:
            element = list.getList()[0]
            if element is None and not element.isNull():
                print("Received malformed Detach header: handle can't be null")
            self.handle = unwrapper.unwrapUInt(element)
        if size > 1:
            element = list.getList()[1]
            if element is not None and not element.isNull():
                self.closed = unwrapper.unwrapBool(element)
        if size > 2:
            element = list.getList()[2]
            if element is not None and not element.isNull():
                code = element.getCode()
                if code not in (AMQPType.amqpType.getValueByKey('LIST_0'),AMQPType.amqpType.getValueByKey('LIST_8'),AMQPType.amqpType.getValueByKey('LIST_32')):
                    print("Expected type 'ERROR' - received: " + str(element.getCode()))
                self.error = AMQPError.amqpError(None, None, None)
                self.error.fromArgumentsList(element)


    def toString(self):
        return "AMQPDetach [handle=" + str(self.handle) + ", closed=" + str(self.closed) + ", error=" + str(self.error) + ", code=" + str(self.code) + ", doff=" + str(self.doff) + ", type=" + str(self.type) + ", channel=" + str(self.channel) + "]"

    def setHandle(self, handle):
        self.handle = handle

    def getHandle(self):
        return self.handle

    def setClosed(self, closed):
        self.closed = closed

    def getClosed(self):
        return self.closed

    def setError(self, error):
        self.error = error

    def getError(self):
        return self.error

File: IoTbrokerClient/src/iot-protocols/amqp/classes/headeramqp/amqpdetach.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqpdetach.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvfixed.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvnull.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"

AMQPDetach::AMQPDetach() : AMQPHeader(new AMQPHeaderCode(AMQP_DETACH_HEADER_CODE))
{
    this->handle = NULL;
    this->closed = NULL;
    this->error = NULL;
}

int AMQPDetach::getLength()
{
    int length = 8;
    AMQPTLVList *arguments = this->arguments();
    length += arguments->getLength();
    return length;
}

int AMQPDetach::getType()
{
    return AMQP_DETACH_HEADER_CODE;
}

AMQPTLVList *AMQPDetach::arguments()
{
    AMQPTLVList *list = new AMQPTLVList();

    if (this->handle == NULL) {
        printf("AMQPDetach::arguments");
        return new AMQPTLVList();
    }
    list->addElementWithIndex(0, AMQPWrapper::wrapUInt(AMQPSimpleType::variantToUInt(this->handle)));

    if (this->closed != NULL) {
        list->addElementWithIndex(1, AMQPWrapper::wrapBool(AMQPSimpleType::variantToBool(this->closed)));
    }
    if (this->error != NULL) {
        list->addElementWithIndex(2, this->error->getList());
    }

    unsigned char * data = (unsigned char *) malloc(sizeof(unsigned char *));
    data[0] = this->code->getValue();
    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);

    AMQPType *constructorType = new AMQPType(list->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);

    list->setConstructor(constructor);
    return list;
}

void AMQPDetach::fillArguments(AMQPTLVList *list)
{
    int size = list->getList().size();

    if (size == 0) {
        printf("AMQPDetach::fillArguments::size == 0");
        return;
    }
    if (size > 3) {
        printf("AMQPDetach::fillArguments::wrong size");
        return;
    }
    if (size > 0) {
        TLVAMQP *element = list->getList()[0];
        if (element->isNull()) {
            printf("AMQPDetach::fillArguments::handle");
            return;
        }
        this->handle = AMQPSimpleType::UIntToVariant(AMQPUnwrapper::unwrapUInt(element));
    }
    if (size > 1) {
        TLVAMQP *element = list->getList()[1];
        if (!element->isNull()) {
            this->closed = AMQPSimpleType::boolToVariant(AMQPUnwrapper::unwrapBool(element));
        }
    }
    if (size > 2) {
        TLVAMQP *element = list->getList()[2];
        if (!element->isNull()) {
            if (element->getTypeValue() != AMQP_LIST0_TYPE && element->getTypeValue() != AMQP_LIST8_TYPE && element->getTypeValue() != AMQP_LIST32_TYPE) {
                printf("AMQPDetach::fillArguments::wrong element type");
                return;
            }
            this->error = new AMQPError();
            this->error->fill((AMQPTLVList *)element);
        }
    }
}

JsonVariant *AMQPDetach::getHandle() const
{
    return handle;
}

void AMQPDetach::setHandle(JsonVariant *value)
{
    handle = value;
}

JsonVariant *AMQPDetach::getClosed() const
{
    return closed;
}

void AMQPDetach::setClosed(JsonVariant *value)
{
    closed = value;
}

AMQPError *AMQPDetach::getError() const
{
    return error;
}

void AMQPDetach::setError(AMQPError *value)
{
    error = value;
}
